Ol ccwtmle

olivierlabayle opened the pull request:
Hi @joshua-slaughter @SjoerdVBeentjes,

In this draft PR there are:

- Some code cleaning (unused functions etc...). 
- When looking at [this plot](https://targene.github.io/TMLE.jl/previews/PR138/examples/case_control_experiment/), something I noticed is that the error bars are much smaller than the classic TMLE. I am not sure this is expected? To contrast, I have made a different implementation for the point estimate and the gradient in the CCW case that is a straightforward implementation of the papers' description (I believe). Now looking at the plot I get with this new implementation, it looks like the error bars are much wider. It would be good if we could develop an understanding of why that is. Maybe by contrasting your gradient and estimate from `ccw_cluster_ic(IC_full::AbstractVector, y::AbstractVector, q0::Float64)` and my gradient and estimate from `gradient_and_estimate(ct_aggregate, gradient_Y_X, y, weights)`. 

I've tried to keep all commits unitful to help you referencing changes if necessary. If you could have a look that would be great, we can meet to discuss this further when you've had a look. Let me know.

<img width="769" height="470" alt="image" src="https://github.com/user-attachments/assets/78a69a1a-8ecc-47be-938d-765795fc3582" />

joshua-slaughter commented on src/counterfactual_mean_based/estimators.jl:
```diff
@@ -110,7 +110,7 @@ function (tmle::Tmle)(Ψ::StatisticalCMCompositeEstimand, dataset; cache=Dict(),
     nomissing_dataset = nomissing(dataset, variables(relevant_factors))
     # If prevalence is provided, we need to ensure we have a stable experimental unit (J controls per case)
     !isnothing(tmle.prevalence) ? nomissing_dataset = get_matched_controls(nomissing_dataset, relevant_factors) : nomissing_dataset
```
Thoughts on the matched controls functionality? I recall you were skeptical about this earlier.

joshua-slaughter commented on test/counterfactual_mean_based/fluctuation.jl:
```diff
@@ -199,5 +196,24 @@ end
     @test mean(gradient) ≈ 0.0 atol=1e-4
 end
 
+@testset "Test TMLE.gradient_and_estimate: cqse control weighting" begin
```
Nice test, small typo in "cqse"

joshua-slaughter commented:
Hey @SjoerdVBeentjes, the difference is that in Olivier implementation the error bars are wider, however, this is because he had accidentally overwritten the correct CCW gradient with the canonical TMLE gradient. After removing the error both approaches agree.

olivierlabayle commented:
I have now:

- Fixed the dead code and both this implementation and Josh's give the same results. This implementation is slower (~4x) but clearer and requires less changes to the codebase. Since this is not really a speed limiting operation (~ms for 500 000 samples) I think I would favour clarity here but happy to discuss.
- I have slightly modified the `get_matched_controls` function to be a "no-operator" when the dataset already contains exactly J controls per samples and provide some logs otherwise.
- I have tried to give a little more context in the example made by Josh since this is meant as a tutorial. Feel free to improve/modify it after we merge this branch into yours.
- Added some further refactoring of the code

My final question about why the variance of the classic TMLE and CCW-TMLE seem to go in opposite direction with increased sampling prevalence remains. Maybe @SjoerdVBeentjes has an explanation for it?

Please have a final review before we merge this in to yours, then the merge into main should be pretty quick depending on whether you want to revise your tutorial a bit.

olivierlabayle commented:
Post thoughts regarding the variance:

For a given prevalence, the "actual" sample size for the CCW TMLE is the number of cases which is low when q is low and increases when q increases. This means that the variance of the CCW will decrease with q since we are keeping the "total" sample size fixed (100 000). Still not sure why the classic TMLE has low variance at small q though...
